### PR TITLE
improve constraint coverage tests

### PR DESCRIPTION
**Pull request by wandmagic**
# Committer Notes

nothing crazy here just an improvement on detecting constraints and assuring coverage

anything with an ID in a constraint file is considered a constraint

and each constraint gets its own "scenario" so we can get a red or green check if its covered or not

also allows npm run:coverage to test coverage and npm run test:constraints to run each part off the test suite independently

### All Submissions:

- [ ] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [ ] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [ ] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

File: features/steps/fedramp_extensions_steps.ts [hunk 1]
```diff
@@ -437,10 +425,11 @@ When(
       constraintIds = [...constraintIds, ...constraints];
     }
     constraintIds = [...new Set(constraintIds)].sort();
+
+    console.log(`Extracted ${constraintIds.length} unique constraint IDs`);
```
Comment (david-waltermire) [inline features/steps/fedramp_extensions_steps.ts:429]: This looks like a duplicate log message.

File: features/steps/fedramp_extensions_steps.ts [hunk 2]
```diff
@@ -611,3 +592,26 @@ When("I analyze the YAML test files for each constraint ID", function () {
...
+
+// New step definition for the "Ensuring full test coverage for "<constraint_id>"" scenario
+Then("I should have both FAIL and PASS tests for constraint ID {string}", function (constraintId) {
+  const testCoverage = testResults[constraintId];
+
+  if (!testCoverage) {
+    console.log(`${constraintId}: No tests found`);
+    expect.fail(`Constraint ${constraintId} has no tests`);
+  } else if (!testCoverage.pass) {
+    console.log(`${constraintId}: Missing positive test`);
```
Comment (david-waltermire) [inline features/steps/fedramp_extensions_steps.ts:604]:
```suggestion
    console.log(`${constraintId}: Missing at least one positive test`);
```

File: features/steps/fedramp_extensions_steps.ts [hunk 3]
```diff
@@ -611,3 +592,26 @@ When("I analyze the YAML test files for each constraint ID", function () {
...
+    expect.fail(`Constraint ${constraintId} is missing a positive test`);
+  } else if (!testCoverage.fail) {
+    console.log(`${constraintId}: Missing negative test`);
+    expect.fail(`Constraint ${constraintId} is missing a negative test`);
+  } else {
+    console.log(`${constraintId}: Fully covered`);
```
Comment (david-waltermire) [inline features/steps/fedramp_extensions_steps.ts:610]:
```suggestion
    console.log(`${constraintId}: Has minimal required coverage`);
```

File: package.json [hunk 4]
```diff
@@ -8,6 +8,8 @@
     "federalist": "make init-repo && npm run build:validation-ui && npm run link:validation-ui",
     "link:validation-ui": "ln -sf ./src/web/dist _site",
     "test": "cross-env NODE_OPTIONS=\"--loader ts-node/esm --no-warnings --experimental-specifier-resolution=node\" cucumber-js",
+    "test:constraints": "cross-env NODE_OPTIONS=\"--loader ts-node/esm --no-warnings --experimental-specifier-resolution=node\" cucumber-js --tags @constraints",
+    "test:coverage": "cross-env NODE_OPTIONS=\"--loader ts-node/esm --no-warnings --experimental-specifier-resolution=node\" cucumber-js --tags @full-coverage",
```
Comment (david-waltermire) [inline package.json:12]: Is this setup to run on `make test`?

Comment (wandmagic) [reply package.json:12]: make test is setup to run both 
build-validations:
@npm run test

File: features/steps/fedramp_extensions_steps.ts (hunk 1 above)
Comment (wandmagic) [reply features/steps/fedramp_extensions_steps.ts:429]: removed

**Review by david-waltermire — APPROVED** [review]

**Review by Rene2mt — APPROVED** [review]

**Pull request merged**